rust/primary/src/header_waiter.rs

```diff
@@ -0,0 +1,292 @@
+use crate::error::{DagError, DagResult};
+use crate::messages::Header;
+use crate::primary::{PrimaryMessage, PrimaryWorkerMessage, Round};
+use bytes::Bytes;
+use config::{Committee, WorkerId};
+use crypto::{Digest, PublicKey};
+use futures::future::try_join_all;
+use futures::stream::futures_unordered::FuturesUnordered;
+use futures::stream::StreamExt as _;
+use log::{debug, error};
+use network::SimpleSender;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use store::Store;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::time::{sleep, Duration, Instant};
+
+/// The resolution of the timer that checks whether we received replies to our sync requests, and triggers
+/// new sync requests if we didn't.
+const TIMER_RESOLUTION: u64 = 1_000;
+
+/// The commands that can be sent to the `Waiter`.
+#[derive(Debug)]
+pub enum WaiterMessage {
+    SyncBatches(HashMap<Digest, WorkerId>, Header),
+    SyncParents(Vec<Digest>, Header),
+}
+
+/// Waits for missing parent certificates and batches' digests.
+pub struct HeaderWaiter {
+    /// The name of this authority.
+    name: PublicKey,
+    /// The committee information.
+    committee: Committee,
+    /// The persistent storage.
+    store: Store,
+    /// The current consensus round (used for cleanup).
+    consensus_round: Arc<AtomicU64>,
+    /// The depth of the garbage collector.
+    gc_depth: Round,
+    /// The delay to wait before re-trying sync requests.
+    sync_retry_delay: u64,
+    /// Determine with how many nodes to sync when re-trying to send sync-request.
+    sync_retry_nodes: usize,
+
+    /// Receives sync commands from the `Synchronizer`.
+    rx_synchronizer: Receiver<WaiterMessage>,
+    /// Loops back to the core headers for which we got all parents and batches.
+    tx_core: Sender<Header>,
+
+    /// Network driver allowing to send messages.
+    network: SimpleSender,
+    /// Keeps the digests of the all certificates for which we sent a sync request,
+    /// along with a timestamp (`u128`) indicating when we sent the request.
+    parent_requests: HashMap<Digest, (Round, u128)>,
+    /// Keeps the digests of the all tx batches for which we sent a sync request,
+    /// similarly to `header_requests`.
+    batch_requests: HashMap<Digest, Round>,
+    /// List of digests (either certificates, headers or tx batch) that are waiting
+    /// to be processed. Their processing will resume when we get all their dependencies.
+    pending: HashMap<Digest, (Round, Sender<()>)>,
+}
+
+impl HeaderWaiter {
```

Comment:
Can we get a bad header and try to sync invalid batches/certificates?

Reply:
yes it may happen, I don't see a way to avoid that. The current thinking is that we try to sync anyway, and will eventually cleanup (thus top synching) upon GC.